### DISCUSSION TITLE
how to refer inner cross reference

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-langium/langium#1468] [Q&A]
FunctionBlock:
    'FUNCTION_BLOCK'  name=Identifier
    (methods+=Methods)*
    'END_FUNCTION_BLOCK'
;

Methods:
    'METHOD'  name=Identifier  
     'END_METHOD'
;

if i want to refer function method name at below rule ,how should i refer it

aa:
  fbMethodName=[FunctionBlock:?]

**Comment by cdietrich** [comment]
depends on your usecase? your impl of the scoping etc

`fbMethodName=[FunctionBlock:Identifier]`

or

```
QualifiedName: Identifier ("." Identifier)*;
...
fbMethodName=[FunctionBlock:QualifiedName]
```

**Reply by zrwsmd** [reply]
if i write like this fbMethodName=[FunctionBlock:Identifier],i think it probably refer to function block name instead of function block method name

**Reply by cdietrich** [reply]
you always write `propertyName=[TypeYouWantToRefer:RuleYouWantToParseForIt]`

**Reply by cdietrich** [reply]
so e.g `fbMethodName=[Methods:QualifiedName]` or `fbMethodName=[Methods:Identifier]`

**Reply by zrwsmd** [reply]
i have try this,but seems did not work out well,this method is defined in function block ,so it should bound with current function block

**Reply by cdietrich** [reply]
maybe you can be more explicit. as i said

> your impl of the scoping etc

so how did you implement scoping

**Reply by zrwsmd** [reply]
for this requirement,how should i implement scoping,offical demo is a little abstract,

**Reply by cdietrich** [reply]
Can you please give a complete example

**Reply by zrwsmd** [reply]
[demo.txt](https://github.com/eclipse-langium/langium/files/15090415/demo.txt)

**Reply by zrwsmd** [reply]
below is corresponding example


TYPE
	electrical_struct: STRUCT
	 productId: INT;
	 vendor: STRING;
	 voltage: LREAL := 3600.589;
	 hasBrake: BOOL;
     
   END_STRUCT
END_TYPE


FUNCTION_BLOCK TemperatureConverter
VAR
     ccc: R_TRIG; // 定义一个实数类型的变量 celsius，用于存储摄氏温度
    dd: REAL;
    ee:LINT;
    ff:BOOL;
   

END_VAR



METHOD ConvertCelsiusToFahrenheit(celsiusTemp: REAL): REAL

END_METHOD

END_FUNCTION_BLOCK



PROGRAM PLC_PRG
VAR
    elec: electrical_struct;
    state: INT:=99;
	tempConverter: TemperatureConverter; 
END_VAR
 voltage := tempConverter.ConvertCelsiusToFahrenheit(56); // 将摄氏温度转换为华氏温度

**Reply by cdietrich** [reply]
can you also provide a complete grammar

**Reply by cdietrich** [reply]
e.g how did you implement your dot expressions

**Reply by zrwsmd** [reply]
what grammar,you mean this?
[grammar.txt](https://github.com/eclipse-langium/langium/files/15090564/grammar.txt)

**Reply by cdietrich** [reply]
`voltage := tempConverter.ConvertCelsiusToFahrenheit(56); `

what kind of construct is `tempConverter.ConvertCelsiusToFahrenheit`

**Reply by cdietrich** [reply]
also a textual grammar (the langium file), minified (just the problem usecase, nothing else) would help

**Reply by zrwsmd** [reply]
i want to validate function block method name,if this method  name did not exist in this function block ,it should report an error
![image](https://github.com/eclipse-langium/langium/assets/26995040/9faa2920-0d4c-40b1-93cc-18d28ebadd80)

**Reply by cdietrich** [reply]
again:
what kind of construct is a.b.
is this one reference or two

**Reply by zrwsmd** [reply]
it is a method call,tempConverter is declared in 
![image](https://github.com/eclipse-langium/langium/assets/26995040/c8e79380-a38a-44ce-8fc6-a16962986bea)

**Reply by cdietrich** [reply]
again is this one reference to tempConverter or a second reference to ConvertCelsiusToFahrenheit
or just a reference to ConvertCelsiusToFahrenheit under the name "tempConverter.ConvertCelsiusToFahrenheit"

**Reply by cdietrich** [reply]
e.g. in typescript it would be two references

**Reply by zrwsmd** [reply]
the connection to github is not well。you can analogy 'FUNCTION_BLOCK' as 'class' in typescript or c++.it define a series of methods which we can call,it should be two references

**Reply by cdietrich** [reply]
so then in the scoping of the second reference you should ask the first reference for its methods

**Reply by zrwsmd** [reply]
i guess so

**Reply by cdietrich** [reply]
then you should give that a try

**Reply by zrwsmd** [reply]
I don't know how to write without thinking

**Reply by cdietrich** [reply]
maybe this terribly old example gives some idea
https://github.com/cdietrich/CoreDsl-Langium/blob/127dba30229639066022327d79f6a590ece9532e/src/language-server/core-dsl-scope-computation.ts#L30

**Reply by zrwsmd** [reply]
thanks a lot,actually ,i am confused about langium lifecycle,the offical document is a bit difficult to understand

**Reply by cdietrich** [reply]
the concept for if there is a reference someRef=[Xxxxxx] and you write Yyyyyyy
and langium tries to figure out what Xxxxxxx instance Yyyyyyy is is called scoping
and implemented in the scope provider.

maybe https://eclipse.dev/Xtext/documentation/303_runtime_concepts.html#scoping also gives some inspiration

**Reply by zrwsmd** [reply]
thanks a lot

**Comment by msujew** [comment]
Hey @zrwsmd, I would recommend taking a look at [this scoping guide](https://langium.org/guides/scoping/class-member/). That is probably what you're looking for.

**Reply by zrwsmd** [reply]
thanks a lot

**Reply by zrwsmd** [reply]
in this langium-lox project example .i define a function name is SuperClass and property 'a' ,also another function name is SuperClass666 and property 'y',when i construct a SuperClass666 instance ,it seems did not validate property,that what i want to achieve
![image](https://github.com/eclipse-langium/langium/assets/26995040/9fce65a1-8d03-4b43-ad68-abaaee5c7a46)

**Reply by cdietrich** [reply]
in the playground you dont implement scoping

**Reply by zrwsmd** [reply]
you mean define a class  extends DefaultScopeProvider,right?

**Reply by cdietrich** [reply]
by default all objects are visible in a reference. if you want to restrict that you need to customize DefaultScopeProvider yes.

**Reply by zrwsmd** [reply]
langium-lox project example need to define variable 'abc' outside Class,then it can be refered,my requirement is  like define a variable inside a Class(can not define outside Class),then i instance a class and refer it variable(e.g  abc) .how should i write grammar rule

![image](https://github.com/eclipse-langium/langium/assets/26995040/819ae9f5-cb80-4cad-a647-0391720132d3)

**Reply by cdietrich** [reply]
VariableDeclaration:
name=ID ":" type=Type

**Reply by zrwsmd** [reply]
so if i want to validate this variable 'abc' belong to Class SubClass, instead of other Class, should i  also  define a class extends DefaultScopeProvider?

**Reply by cdietrich** [reply]
yes.

there you need to get the x. determine xes type. then ask the type for its variables (this is done here: https://github.com/TypeFox/langium-lox/blob/a821a08b967e1eb260079ca6e4a612a1c46000e9/src/language-server/lox-scope.ts#L24)

**Reply by cdietrich** [reply]
i am also not sure why you man by "validate"
scoping determine which objects are visible at a cross reference place

**Reply by zrwsmd** [reply]
i have noticed this ,but not clearly  catch its meaning,how can i debug into lox-scope.ts,

**Reply by cdietrich** [reply]
simply setting a breakpoint. launch the "run extension" host. launch the "attach to language server" should work.

**Reply by zrwsmd** [reply]
why did this method createScopeForNodes simply have memeber params ,how could it bound with corresponding Class or what is this method do
![image](https://github.com/eclipse-langium/langium/assets/26995040/11000ca4-a1dd-413e-83c6-f363adcca451)
![image](https://github.com/eclipse-langium/langium/assets/26995040/00775cb7-6fe0-4dac-94a3-6d87376bf9e8)

**Reply by cdietrich** [reply]
Without looking at the code it will collect the member off the  class and those of all superclasses

**Reply by zrwsmd** [reply]
i thinks so,but still did not get what is the meaning of this method  `createScopeForNodes`,expand its scope?

**Reply by cdietrich** [reply]
See https://github.com/eclipse-langium/langium/pull/665

in turns a stream of nodes into a scope object

**Reply by zrwsmd** [reply]
so what is difference with this StreamScope and global scope

**Reply by cdietrich** [reply]
I don’t understand this question 

global scope is I don’t want to decide which objects should be visible

**Reply by msujew** [reply]
The `StreamScope` is an implementation of the `Scope` interface. The global scope is just an instance of the `Scope` interface that represents all globally available objects.

**Reply by zrwsmd** [reply]
in which condition should i override DefaultScopeProvider&getScope,
in which condition should i override DefaultScopeProvider&getGlobalScope,

**Reply by msujew** [reply]
You want to override `getScope` in case you want to adjust how a **specific reference** is resolved. `getGlobalScope` should be overriden to determine which elements are available for referencing on a global level.

**Reply by zrwsmd** [reply]
can you please describe it in a more concise and understandable way or list a simple example to describe it ,a little abstract

**Reply by msujew** [reply]
Every concise way of explaining how scopes work will be pretty abstract. If you want a better understanding of how scoping works, I would recommend to read a book on Xtext. The authors explain scoping pretty in depth. See [Markus Völter's *DSL Engineering*](http://voelter.de/data/books/markusvoelter-dslengineering-1.0.pdf). Starting at page 230, he explains how scoping and linking in Xtext work. These mechanism are very similar to how they work in Langium.
